Add `missing_transmute_annotations` lint

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust-clippy/issues/715.

r? @blyxyas 

changelog: Add `missing_transmute_annotations` lint

GuillaumeGomez commented:
And as usual I discover that there is already a `transmute` module which does pretty much all I originally did manually. Moved the lint there.

blyxyas commented:
#715?? Guillaume, that issue was open 8 years ago. Really going on a journey back inntime

GuillaumeGomez commented:
Which also provides type suggestions. Maybe I should add that.

GuillaumeGomez commented:
> #715?? Guillaume, that issue was open 8 years ago. Really going on a journey back inntime

Told you I was going through issues. :p

I generally pick a random page and start taking a look there.

> There's also #11047 for basically an identical lint, it's inactive though

Didn't see that one. ^^' Oh well, since it's inactive I think it's fine. I added suggestions as well.

blyxyas commented:
I'll have to reroll this one, we are at Feb. 9 and I have practically no progress on performance (which is kinda really important), I've also set myself on vacation for this exact reason (I'd love to not have my busyness status in version control D:).

r? rust-lang/clippy

y21 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,35 @@
+            MISSING_TRANSMUTE_ANNOTATIONS,
+            last.ident.span.with_hi(path.span.hi()),
+            "transmute used without annotations",
+            "consider adding missing annotations with `transmute::<IN, OUT>`",
```
```suggestion
            "consider adding missing annotations",
```
We can shorten it since this is just repeating the suggestion :)

GuillaumeGomez replied:
:+1:

y21 commented on tests/ui/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,17 @@
+        //~^ ERROR: transmute used without annotations
+
+        // Should not warn.
+        let x = std::mem::transmute::<[u16; 2], i32>([1u16, 2u16]);
```
Can we also have a test for a bad `transmute` coming from an external macro?

GuillaumeGomez replied:
I added a test for macros as well.

y21 replied:
I meant a macro from a different crate :)
Local macros are fine to lint (like it's doing right now), but not external macros, since the user can't do much about them.

GuillaumeGomez replied:
Added as well.

y21 reviewed (commented):
Looks like a really nice lint to have! Implementation looks pretty good already. Not sure about the category and if we want to have it be warn-by-default though - that might deserve discussion :)

GuillaumeGomez commented on clippy_lints/src/transmute/mod.rs:
````diff
@@ -520,6 +521,32 @@ declare_clippy_lint! {
+    /// ```
+    #[clippy::version = "1.77.0"]
+    pub MISSING_TRANSMUTE_ANNOTATIONS,
+    suspicious,
````
I agree with your analysis. Even if the `transmute` as the last statement/tail expression, it doesn't hurt to have the types information right on the `transmute` without needing to go back up to the function declaration (for long functions). And if there is a change in the returned type, it'll force users to come back to this code to ensure it's still what they want.

y21 replied:
Ah, I just looked at the other PR, and saw that @xFrednet you had a similar concern here: https://github.com/rust-lang/rust-clippy/pull/11047#discussion_r1252221990 , do you still think it should be allow-by-default?

I'm of the opinion that this would generally be useful to have, as warn-by-default, since the advantage of protecting against type inference for such a heavy hammer (`transmute`) IMO outweigh the occasional situation-dependent `#[allow]` one might need. I'm interested in hearing others' opinions, tho.

xFrednet replied:
This feels like a lint that should be warn-by-default. Though, we should run a lintcheck to see how many reports it generates. If it's too many, then it might be better to have it in `pedantic`. If we generate a lot of new warnings, it might be counterproductive and cause users to just allow the lint without thinking about the consequences.

GuillaumeGomez replied:
I don't think "it generates too many warnings" is a good point to not keep it as warn-by-default, quite the opposite. This lint isn't emitted if it's generated from an external macro or if the result of the `transmute` is used directly in a let binding with type annotation. Anything else should most definitely warn as it could allow to uncover some problematic code.

xFrednet replied:
> I don't think "it generates too many warnings" is a good point to not keep it as warn-by-default

It can be hard to pick the correct lint group/level, it depends on the lint and what it detects. As said, just by reading this, I would consider it to be a warn-by-default lint. Running lintcheck is meant to confirm this feeling.

I'd still argue that the number of warnings should be a factor or at least be looked at for mainly two reasons:

1. If there are too many warnings, it's an indication that the suggested change might not be idiomatic.
2. We have to think about the users. If Clippy generates too many errors/warnings, users are more prune to just disabling a lint and being annoyed. There were a few lints, I would have considered being warn-by-default, but the community disagreed with. (Which is totally fine, discussions on these topics are important) I believe that some of these cases were not because of the lint, but because of the amount of required changes.

---

Generally speaking and slightly off-topic, I would hope that most projects enable the `pedantic` group as well. For small hacking scrips it might be overkill, but on bigger projects I found it to be a big help.

GuillaumeGomez replied:
\o/

y21 replied:
Out of curiosity: how many crates was it checking? I also ran it earlier and got 16 hits, but that was with 200 crates. (Not that it changes my view on the category; I still agree)

GuillaumeGomez replied:
26 as far as I can see.

xFrednet replied:
@y21 I think it might be worth it to update the default list in this repo to include a larger set like yours.

y21 replied:
Yeah, I agree. I think the default list is often not large enough to come to useful conclusions, there has been instances where the default list hasn't found anything but 200 found a bunch. But it also requires some manual reviewing if we do that I think (instead of blindly using lintcheck's `popular-crates` subcommand). For example, in one crate, there are *200,000* hits for the match_same_arm lint.

GuillaumeGomez commented on clippy_lints/src/transmute/mod.rs:
```diff
@@ -520,6 +521,32 @@ declare_clippy_lint! {
+    ///
+    /// ### Why is this bad?
+    /// If not set, some unexpected output type could be retrieved instead of the expected one,
+    /// potentially leading to invalid code.
```
Included your suggestion. :)

GuillaumeGomez commented:
Applied suggestions.

Centri3 commented on tests/ui/missing_transmute_annotations.stderr:
```diff
@@ -0,0 +1,40 @@
+error: transmute used without annotations
+  --> $DIR/missing_transmute_annotations.rs:17:32
+   |
+LL |         let x: i32 = std::mem::transmute::<[u16; 2], _>([1u16, 2u16]);
```
`_` really shouldn't be linted, you're explicitly saying you want it to be inferred

GuillaumeGomez replied:
Indeed, I implemented that.

y21 replied:
Isn't this still linted?
https://github.com/rust-lang/rust-clippy/blob/235c9f397f36f3c89d7122851776d8421ccf8fcc/tests/ui/missing_transmute_annotations.stderr#L22-L26

GuillaumeGomez replied:
Unless I missed something, there is no `let` binding with type annotation on this line.

y21 replied:
Ah, I might've misunderstood @Centri3's comment. The way I understood it was that it should avoid linting entirely when `_` is specified no matter what, whereas the current impl allows inferred types when there is an enclosing type-annotated binding.

If I did understand it correctly, I'm not sure if it's worth special casing explicit `_` in the generic arguments and treating them differently like that (which it's currently not doing anyway). All the same caveats and inference arguments still apply when writing `transmute::<_, _>()` as opposed to just `transmute()`.

Thinking about it, I do like the current heuristics (checking if the parent node is a local variable with a type annotation).
It might make sense to actually extend this to also check if this is the last expression in the function and not lint then? I mentioned this in the other comment about the lint category and IMHO it would rule out a lot of noise.
I did a lintcheck run and almost all of the hits were small functions where it's just a transmute and the target type is already annotated in the function signature.

We could also have a configuration for a 'stricter' mode, to not even allow these cases, which I think is a good idea if we're gonna leave it as a warn-by-default lint. WDYT?

Centri3 replied:
I wasn't thinking about the `let` at all, I think it shouldn't lint it with `transmute::<_, _>()`. I can't think of any reasons to actually want it inferred, but better safe than sorry. A stricter mode should definitely be added if it's warn-by-default, and maybe if it's allow-by-default too

GuillaumeGomez replied:
> It might make sense to actually extend this to also check if this is the last expression in the function and not lint then? 

We already discussed about it in this PR and it was actually one of the motivations for this PR: changing the returned type of a function can alter the `transmute` returned value and that is likely not what the user would want.

@Centri3: I don't think it's a good idea. Having `transmute::<_, _>(` is the same as having `transmute(`. It brings no additional information and should be linted against.

Centri3 replied:
It does give more information, though: you want it to be inferred. Why's that? No idea. If this is allow-by-default clearly somebody wants to apply the lint's suggestion, and won't do this just to silence clippy.

And playing the devil's advocate, you can't always specify a generic, especially so if it isn't even generic. Say in macro generated code, or just generated code in general. This is absolutely reckless but if you're using transmute, you hopefully know what you're doing anyway, and any preconditions are upheld elsewhere. At the end of the day you can just `#[allow]` it, but that wouldn't be linted if you switch to the stricter mode (say you're phasing this out of your codebase, or something, I dunno)

GuillaumeGomez replied:
Is there a way to check how many crates would be impacted by this lint and advise from that? Imo, if we don't emit the lint if `_` is used of any of the generics, then the lint loses a big part of its use.

Centri3 replied:
Try running lintcheck and only lint if `_` is used, I suppose. I don't really mind if this is denied by default so that those cases (where somebody uses `_` erroneously) are still linted. This could be tied to a configuration option

I do agree with you that those cases are important, now that I think about it.

y21 replied:
> We already discussed about it in this PR and it was actually one of the motivations for this PR: changing the returned type of a function can alter the transmute returned value and that is likely not what the user would want.

For long functions where you might not be able to immediately spot the type annotation in the signature, I agree. But for functions where the transmute call is the only expression in the function and there are no other statements, I'm not so sure after thinking about it a bit more.
For example, it seems excessive to have to write
```rs
fn i32_to_ne_bytes(val: i32) -> [u8; 4] {
  unsafe { std::mem::transmute::<i32, [u8; 4]>(val) }
}
```
Arguably, this is just as local as a `let` binding with type annotations

GuillaumeGomez replied:
Whatever the length of the function, it's easy to forget to update both the returned type of the function and eventually check the `transmute`.

y21 replied:
I'm still not entirely sure here. I opened a Zulip stream for this, just to get some more opinions: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/missing_transmute_annotations

y21 replied:
Looks like the "majority" on the zulip thread was of the opinion that we should lint this, so let's go with that, then.

I also hadn't considered that people might fold functions in their IDEs such that the transmute is not as apparent, so that's another point for why we should lint this anyway

bors commented:
:umbrella: The latest upstream changes (presumably #12040) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Updated, cleaned up the commit history as well and implemented the two limitations:
 * not emitting if it comes from an external macro
 * not emitting if the `transmute` returned value is put into a let binding with type annotations

GuillaumeGomez commented:
Anything else to be done here?

Centri3 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,53 @@
+        last.ident.span.with_hi(path.span.hi()),
+        "transmute used without annotations",
+        "consider adding missing annotations",
+        format!("{}::<{from_ty}, {to_ty}>", last.ident.as_str()),
```
If one/more of the types can be left inferred, can we keep it like that here? e.g., `let _: T`.

GuillaumeGomez replied:
I'd rather not.

Centri3 replied:
Why not? Is it because it'd require rewriting it? I don't see why this shouldn't be done.

Centri3 replied:
What I meant was that, if it's within a `let` statement for example, we could keep `Dst` as `_`. If it's better to provide both directly, and this is still an issue with `let`, then we shouldn't special case `let` at all, no?

I see what you mean that providing them is better, but special-casing it then not suggesting it is a bit contradictory.

GuillaumeGomez replied:
The idea behind my position is that if you ever remove the `let` binding or change the type annotation, if we originally suggested both types, it'll prevent compilation and require user to check that the code is still valid. I think it's pretty important considering how easy it is to have unexpected behaviours with `transmute`.

Centri3 commented on tests/ui/missing_transmute_annotations.rs:
Please add tests for ADTs.
Also, `super` can be a problem with paths I think. #10854 and here's the fix #11370. If it's a problem, you don't have to fix it, but a FIXME would be nice.

GuillaumeGomez replied:
Good idea.

Centri3 commented on tests/ui/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,64 @@
+
+macro_rules! local_bad_transmute {
+    ($e:expr) => {
+        std::mem::transmute($e)
```
Can you special case `_` here in all macros? The type can't always be specified like this automatically

GuillaumeGomez replied:
In such cases, I think it means the macro should be rewritten instead of us dropping the level of detection.

Centri3 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,53 @@
+    if !missing_generic {
+        return false;
+    }
+    // If it's being set as a local variable value...
```
We should include closures here imo, if they have type annotations. This could include the parameter if it's specified as well, but not something like:
```rust
// i64 is known, but [i32; 2] is never specified
|x: i32, y: i32| -> i64 unsafe { transmute::<[i32; 2], _>([x, y]) };
```

GuillaumeGomez replied:
I don't think we should special case closures either. It adds a lot of complexity to detect such cases for a very marginal gain imo.

Centri3 replied:
Not really. The return type part is easy, and the parameter is too - Get the `Res` from the `QPath`, the `HirId` of the local, then the parent's parent *should* include the types in `FnDecl`.

I wouldn't block it on this so if it's not done I can add it after, eventually.

GuillaumeGomez replied:
Not in this sense. I meant, for the same reason we want to enforce having type annotations when `transmute` is used as return expr, we want to enforce for it for closures as well.

Centri3 replied:
I don't agree with that as closures are usually private and not public. It's more akin to `let` than functions. I don't think you can fold a closure either in IDEs.

GuillaumeGomez replied:
I don't see why closures being private or public changes anything. If you want to bring this back to debate, let's continue on zulip. ;)

Centri3 replied:
I'll bring up my current points on zulip sometime.

GuillaumeGomez replied:
:+1: There is an open thread already, don't hesitate to comment there directly. ;)

y21 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,53 @@
+    if !missing_generic {
+        return false;
+    }
+    // If it's being set as a local variable value...
```
This doesn't currently check that only the output type parameter was `_`. So for example this is not linted
```rs
let _: i64 = transmute([x, y]);
```
despite the missing input type annotation.

y21 replied:
It should presumably also look past ancestor blocks, so that `let _: T = unsafe { transmute::<_, _>() };` is linted. The parent expression of transmute is the unsafe block, not the let binding.

y21 replied:
Though I suppose allowing blocks might be getting close to the closure case pointed out by Centri, or other single-expression functions.
It does feel a bit arbitrary to allow specific cases like these but forbid closures/blocks though.

y21 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,53 @@
+    expr_hir_id: HirId,
+) -> bool {
+    let last = path.segments.last().unwrap();
+    if in_external_macro(cx.tcx.sess, last.ident.span) {
```
Makes me wonder if this shouldn't just apply to *all* transmute lints, much like how `methods/mod.rs` also blanket allows method calls from external macros. 🤔

Centri3 replied:
`transmute` includes lints that catch UB, so they should include external macros. Like `transmute_null_to_fn` as an example

GuillaumeGomez replied:
I can remove this limitation if you want. No opinion on this so whatever the clippy team prefers.

GuillaumeGomez replied:
Oh. So something outside of this lint. Might be worth opening an issue so it's not forgotten.

GuillaumeGomez commented:
I added check for `let var: _ =`, I also extended the check to look up if the `transmute` call parent is a block that it is assigned in a variable with type annotations (or not). And finally, I added ADT tests as well to check suggestions look correct.

With this I think I didn't forget anything?

GuillaumeGomez commented:
And updated ui test.

bors commented:
:umbrella: The latest upstream changes (presumably #12386) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
As discussed on zulip, if the `transmute` call is the only thing in the function, then the lint won't be emitted.

Anything else to be done?

y21 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,90 @@
+        {
+            return false;
+        }
+    // We check if this transmute is not the only element in the function
+    } else if is_function_block(cx, expr_hir_id) {
```
What happens for something like
```rs
const _: i64 = unsafe { std::mem::transmute([1i32, 1i32]) };
```
Does this get linted? Should this get linted? This is similar to functions, so if we're suppressing it there I want to say no, probably not.
It would be another good test case tho

GuillaumeGomez replied:
It will be linted. I'll add a test case to ensure that.

GuillaumeGomez replied:
Ah no sorry nevermind. For a const block like this, it should not lint. With the update you suggested, it doesn't lint (as expected).

y21 commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,90 @@
+fn is_function_block(cx: &LateContext<'_>, expr_hir_id: HirId) -> bool {
+    for (_, node) in cx.tcx.hir().parent_iter(expr_hir_id) {
+        if let Node::Item(item) = node
+            && let ItemKind::Fn(_, _, body_id) = item.kind
```
`ItemKind::Fn` only applies to free standing functions. This wouldn't work for functions in `impl`s

y21 replied:
This would probably be easier and possibly more accurate with `enclosing_body_owner`. Would also allow us to catch `const` items like in the previous comment

GuillaumeGomez replied:
Noted!

GuillaumeGomez replied:
More like `body_owner` from what I can see from the docs (better to get a `HIrId` rather than a `DefId`).

GuillaumeGomez commented:
Updated!

y21 reviewed (approved):
Looks fine to me. The implementation seems more lax now (e.g. it allows a tail transmute expression in a function no matter how many other statements there are), but that's fine by me and should also make it less controversial.

It would probably be useful to have an optional configuration variable for disallowing inferred transmute node args *everywhere* if a user specifically wants that behavior, but I don't want to block the (initial) PR adding it for too much longer. It could always be improved separately

y21 commented:
The history here is a bit "convoluted", I tried cleaning it up a little bit by resolving my previous comments, but I don't know about @Centri3 's comments. Some of them aren't implemented, but I'm also not sure if you both agreed that its okay to not do it or if there's something left from your side?

GuillaumeGomez commented on clippy_lints/src/transmute/missing_transmute_annotations.rs:
```diff
@@ -0,0 +1,87 @@
+    if let Some(local) = get_parent_local_binding_ty(cx, expr_hir_id) {
+        // ... which does have type annotations.
+        if let Some(ty) = local.ty
+            // If this is a `let x: _ =`, we shouldn't lint.
```
Forgot to update the comment, good catch!

GuillaumeGomez commented:
Fixed comment.

> It would probably be useful to have an optional configuration variable for disallowing inferred transmute node args _everywhere_ if a user specifically wants that behavior, but I don't want to block the (initial) PR adding it for too much longer. It could always be improved separately

Please open an issue and assign me to it so it's not forgotten. :wink:

GuillaumeGomez commented:
> The history here is a bit "convoluted", I tried cleaning it up a little bit by resolving my previous comments, but I don't know about @Centri3 's comments. Some of them aren't implemented, but I'm also not sure if you both agreed that its okay to not do it or if there's something left from your side?

I *think* it was resolved based on [their comment in the zulip conversation](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/missing_transmute_annotations/near/423167535).

Centri3 commented:
I didn't bring up anything, though I don't think it's worth it to block on anything. It was mostly supposed to be about allowing some edge cases (like closures) which don't need to be added, but my rationale behind them.

Everything else is good.

GuillaumeGomez commented:
Thanks for confirming!

@y21 time for r+? :3

y21 commented:
Sounds good, can you squash the commits? `r=me` with that done

@bors delegate+

bors commented:
:v: @GuillaumeGomez, you can now approve this pull request!

If @y21 told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@y21`
<!-- homu: {"type":"Delegated","delegator":"y21","delegate":"GuillaumeGomez"} -->

GuillaumeGomez commented:
@bors r=y21

bors commented:
:pushpin: Commit ee2558223f51b894475dcd1458b21f7ea393508a has been approved by `y21`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=y21 ee2558223f51b894475dcd1458b21f7ea393508a -->
<!-- homu: {"type":"Approved","sha":"ee2558223f51b894475dcd1458b21f7ea393508a","approver":"y21","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit ee2558223f51b894475dcd1458b21f7ea393508a with merge 95c62ffae9bbce793f68a6f1473e3fc24af19bdd...
<!-- homu: {"type":"BuildStarted","head_sha":"ee2558223f51b894475dcd1458b21f7ea393508a","merge_sha":"95c62ffae9bbce793f68a6f1473e3fc24af19bdd"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/8405480314/job/23018273978), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/8405480317/job/23018278597), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/8405480316/job/23018372548)
Approved by: y21
Pushing 95c62ffae9bbce793f68a6f1473e3fc24af19bdd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"y21","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8405480316/job/23018372548","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8405480314/job/23018273978","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/8405480317/job/23018278597"},"merge_sha":"95c62ffae9bbce793f68a6f1473e3fc24af19bdd"} -->
